tools/jacobi_theta_data.cpp

```diff
@@ -0,0 +1,66 @@
+
+#include "mp_t.hpp"
+#include <boost/math/tools/test_data.hpp>
+#include <boost/test/included/prg_exec_monitor.hpp>
+#include <boost/math/special_functions/jacobi_theta.hpp>
+#include <fstream>
```

Comment:
I don't understand what this file is doing. It appears you are spitting out values of the function from your own code; but isn't that tautologous?

Reply:
> @evanmiller : Could you drop this file into `reporting/performance/jacobi_theta_performance.cpp`?

Will do!

Reply:
This is what the other `tools/*_data.cpp` files do – use a high-precision implementation to compare against the low-precision implementations.

Reply:
Sort of, here's how I approach this:

Let's say we have a principal method, plus a bunch of special case handling (asymptotic expansions at one end of the range say as an example).  The main things that can go wrong:

1) The principal method has an outright bug (eg terms in series not correctly calculated).
2) The principal method looses precision due to either cancellation or an incorrect termination condition.
3) There are bugs in the special cases, or the selection logic between them.

We can deal with (1) by comparison to a "known good" - so job done on that I think.

For (2) and (3) what I try and do is calculate spot values at very high precision using the principal method alone, the important thing is to remove all the special case handling as otherwise as @NAThompson says you're not really testing much.  In an ideal world you would use a completely different method, but those are really hard to find.  When I remember, what I should have done (but didn't always) is archive a "naive" version of the principal method in the test value generator file and call that rather than the actual method - I suspect I may have been sloppy and just quickly commented stuff out in the header occasionally though :(

However, with regard to different methods: do the defining series for these functions satisfy that?  They don't have to be fast for test value generation as they're only run the once.  In extreme cases (1F1) I've resorted to interval arithmetic using mpfi_float to handle series with lots of cancellation when generating spot tests.

Reply:
That makes sense. We have essentially two implementations of each function, the `IMAGINARY` path (for `tau` < 1) and the regular path. The code chooses the one that is expected to converge more quickly. So one possibility would be to invert the branch logic so that the test data always chooses the slower path when generating test data.